src/NHibernate/Impl/FutureBatch.cs

```diff
@@ -154,38 +133,14 @@ public object TransformTuple(object[] tuple, string[] aliases)
 
 			public IList TransformList(IList collection)
 			{
-				if (_currentIndex >= _postExecutes.Count)
+				if (_currentIndex >= _batchedQueries.Count)
 					throw new InvalidOperationException(
 						$"Transformer have been called more times ({_currentIndex + 1}) than it has queries to transform.");
 
-				var postExecute = _postExecutes[_currentIndex];
+				var batchedQuery = _batchedQueries[_currentIndex];
 				_currentIndex++;
-				if (postExecute.ExecuteOnEval == null)
-				{
-					return collection;
-				}
-
-				var results = (IList) typeof(List<>)
-					.MakeGenericType(postExecute.ResultType)
-					.GetConstructor(System.Type.EmptyTypes)
-					.Invoke(null);
-
-				if (!postExecute.IsValue)
-				{
-					foreach (var element in (IEnumerable) postExecute.ExecuteOnEval.DynamicInvoke(collection))
-					{
-						results.Add(element);
-					}
-					return results;
-				}
-
-				// When not null on a future value, ExecuteOnEval is fetched with PostExecuteTransformer from
-				// IntermediateHqlTree through ExpressionToHqlTranslationResults, which requires a IQueryable
-				// as input and directly yields the scalar result when the query is scalar.
-				var resultElement = postExecute.ExecuteOnEval.DynamicInvoke(collection.AsQueryable());
-				results.Add(resultElement);
 
-				return results;
+				return batchedQuery.Future?.TransformList(collection);
```

Comment:
In fact I do not think we may have cases where `Future` will be null, but I have not checked all code paths to be sure, so for a patch I have chosen to be on the safe side.
But then we requires here to yield the collection, not `null`.
I will integrate by changing this to:
```c#
return batchedQuery.Future?.TransformList(collection) ?? collection;
```
This may hide/obfuscate a failure case (`TransformList` yielding `null`), but that is a quite unlikely failure.